File: packages/core/src/crypto.ts
/**
 * Key handling: RFC 7638 JWK thumbprints and Ed25519 verification.
 *
 * Everything here runs on WebCrypto, so the same code works on Node, Deno, Bun,
 * and Workers without a native dependency.
 */
import type { webcrypto } from 'node:crypto'

/**
 * An imported public key.
 *
 * Aliased from the Node typings rather than pulling the DOM lib in, which would
 * put `window` and `document` into the type space of a server-side library. The
 * import is type-only and erases completely, so nothing here binds to Node at
 * runtime.
 */
export type PublicKey = webcrypto.CryptoKey

/** A JWK as it appears in a key directory, plus the validity window the Web Bot Auth directory draft adds. */
export interface Jwk {
  readonly kty: string
  readonly crv?: string
  readonly x?: string
  readonly y?: string
  readonly n?: string
  readonly e?: string
  readonly k?: string
  readonly kid?: string
  readonly alg?: string
  readonly use?: string
  readonly key_ops?: readonly string[]
  /** Not valid before, Unix seconds. */
  readonly nbf?: number
  /** Not valid after, Unix seconds. */
  readonly exp?: number
}

export class KeyError extends Error {
  override readonly name = 'KeyError'
}

/** The members RFC 7638 §3.2 includes in the thumbprint, in lexicographic order. */
const THUMBPRINT_MEMBERS: Readonly<Record<string, readonly string[]>> = {
  EC: ['crv', 'kty', 'x', 'y'],
  OKP: ['crv', 'kty', 'x'],
  RSA: ['e', 'kty', 'n'],
  oct: ['k', 'kty'],
}

/**
 * The RFC 7638 canonical form: required members only, lexicographically
 * ordered, no whitespace.
 *
 * Exported because it is the part worth pinning in a test. A thumbprint that
 * disagrees with every other implementation by one byte of JSON is a
 * `key_not_found` verdict that nobody can debug from the outside.
 */
export function canonicalJwkJson(jwk: Jwk): string {
  const members = THUMBPRINT_MEMBERS[jwk.kty]
  if (members === undefined) throw new KeyError(`unsupported key type: ${jwk.kty}`)
  const parts: string[] = []
  for (const member of members) {
    const value = (jwk as unknown as Record<string, unknown>)[member]
    if (typeof value !== 'string' || value === '') {
      throw new KeyError(`key is missing required member "${member}"`)
    }
    parts.push(`${JSON.stringify(member)}:${JSON.stringify(value)}`)
  }
  return `{${parts.join(',')}}`
}

/**
 * RFC 7638 JWK thumbprint, base64url with no padding.
 *
 * This is the `keyid` a Web Bot Auth signer puts in `Signature-Input`. Badge
 * always computes it locally rather than trusting a directory's own `kid`:
 * accepting the directory's label would let one key in a directory impersonate
 * another key in the same directory.
 */
export async function jwkThumbprint(jwk: Jwk): Promise<string> {
  const json = new TextEncoder().encode(canonicalJwkJson(jwk))
  const digest = await crypto.subtle.digest('SHA-256', json)
  return base64UrlEncode(new Uint8Array(digest))
}

/**
 * Members a public JWK may carry. Everything else is dropped.
 *
 * A whitelist rather than a blacklist, for the same reason publishing refuses
 * private members by name: deleting `d` from an exported private key leaves
 * `key_ops: ["sign"]` and `ext: true` behind, which is a public key advertising
 * that it can sign. Enumerating what may be published cannot fail that way.
 */
const PUBLIC_MEMBERS = [
  'kty',
  'crv',
  'x',
  'y',
  'n',
  'e',
  'kid',
  'alg',
  'use',
  'nbf',
  'exp',
] as const

/** Reduce any JWK to the public members a key directory should carry. */
export function toPublicJwk(jwk: Jwk): Jwk {
  const source = jwk as unknown as Record<string, unknown>
  const out: Record<string, unknown> = {}
  for (const member of PUBLIC_MEMBERS) {
    if (source[member] !== undefined) out[member] = source[member]
  }
  return out as unknown as Jwk
}

export function isEd25519(jwk: Jwk): boolean {
  return jwk.kty === 'OKP' && jwk.crv === 'Ed25519'
}

export async function importEd25519PublicKey(jwk: Jwk): Promise<PublicKey> {
  if (!isEd25519(jwk)) throw new KeyError('not an Ed25519 key')
  if (typeof jwk.x !== 'string' || jwk.x === '') throw new KeyError('key is missing "x"')
  return crypto.subtle.importKey(
    'jwk',
    { kty: 'OKP', crv: 'Ed25519', x: jwk.x },
    { name: 'Ed25519' },
    false,
    ['verify'],
  )
}

export async function verifyEd25519(
  key: PublicKey,
  signature: Uint8Array,
  data: Uint8Array,
): Promise<boolean> {
  // A malformed signature makes WebCrypto throw in some runtimes and return
  // false in others. Both mean the same thing here: not verified.
  try {
    return await crypto.subtle.verify({ name: 'Ed25519' }, key, signature, data)
  } catch {
    return false
  }
}

/** Whether a directory key's `nbf`/`exp` window contains `now` (Unix seconds). */
export function keyValidityAt(jwk: Jwk, now: number): 'valid' | 'not-yet-valid' | 'expired' {
  if (typeof jwk.nbf === 'number' && now < jwk.nbf) return 'not-yet-valid'
  if (typeof jwk.exp === 'number' && now >= jwk.exp) return 'expired'
  return 'valid'
}

export function base64UrlEncode(bytes: Uint8Array): string {
  let bin = ''
  for (const b of bytes) bin += String.fromCharCode(b)
  return btoa(bin).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '')
}

export function base64UrlDecode(value: string): Uint8Array {
  if (!/^[A-Za-z0-9_-]*$/.test(value)) throw new KeyError('invalid base64url')
  const padded = value.replace(/-/g, '+').replace(/_/g, '/')
  const bin = atob(padded.padEnd(Math.ceil(padded.length / 4) * 4, '='))
  const bytes = new Uint8Array(bin.length)
  for (let i = 0; i < bin.length; i += 1) bytes[i] = bin.charCodeAt(i)
  return bytes
}

File: sidecar/wba/crypto.go
package wba

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
)

// JWK is a key as published in a directory, plus the validity window the Web
// Bot Auth directory draft adds.
type JWK struct {
	Kty string `json:"kty"`
	Crv string `json:"crv,omitempty"`
	X   string `json:"x,omitempty"`
	Y   string `json:"y,omitempty"`
	N   string `json:"n,omitempty"`
	E   string `json:"e,omitempty"`
	K   string `json:"k,omitempty"`
	Kid string `json:"kid,omitempty"`
	Alg string `json:"alg,omitempty"`
	Use string `json:"use,omitempty"`
	Nbf *int64 `json:"nbf,omitempty"`
	Exp *int64 `json:"exp,omitempty"`
}

// ErrKey reports a key that cannot be used.
var ErrKey = errors.New("unusable key")

// CanonicalJSON returns the RFC 7638 canonical form: required members only, in
// lexicographic order, no whitespace.
//
// Exported because it is the part worth pinning. A thumbprint that disagrees
// with other implementations by one byte of JSON is a key_not_found verdict
// nobody can debug from the outside.
func CanonicalJSON(key JWK) (string, error) {
	quote := func(s string) string {
		encoded, _ := json.Marshal(s)
		return string(encoded)
	}
	need := func(value, member string) error {
		if value == "" {
			return fmt.Errorf("%w: missing required member %q", ErrKey, member)
		}
		return nil
	}

	switch key.Kty {
	case "OKP":
		if err := need(key.Crv, "crv"); err != nil {
			return "", err
		}
		if err := need(key.X, "x"); err != nil {
			return "", err
		}
		return fmt.Sprintf(`{"crv":%s,"kty":"OKP","x":%s}`, quote(key.Crv), quote(key.X)), nil
	case "EC":
		if err := need(key.Crv, "crv"); err != nil {
			return "", err
		}
		if err := need(key.X, "x"); err != nil {
			return "", err
		}
		if err := need(key.Y, "y"); err != nil {
			return "", err
		}
		return fmt.Sprintf(`{"crv":%s,"kty":"EC","x":%s,"y":%s}`, quote(key.Crv), quote(key.X), quote(key.Y)), nil
	case "RSA":
		if err := need(key.E, "e"); err != nil {
			return "", err
		}
		if err := need(key.N, "n"); err != nil {
			return "", err
		}
		return fmt.Sprintf(`{"e":%s,"kty":"RSA","n":%s}`, quote(key.E), quote(key.N)), nil
	case "oct":
		if err := need(key.K, "k"); err != nil {
			return "", err
		}
		return fmt.Sprintf(`{"k":%s,"kty":"oct"}`, quote(key.K)), nil
	}
	return "", fmt.Errorf("%w: unsupported key type %q", ErrKey, key.Kty)
}

// Thumbprint computes the RFC 7638 JWK thumbprint, base64url without padding.
//
// This is the keyid a Web Bot Auth signer presents. It is always computed
// locally rather than read from the directory's own kid member: trusting that
// label would let one key in a directory impersonate another in the same
// directory.
func Thumbprint(key JWK) (string, error) {
	canonical, err := CanonicalJSON(key)
	if err != nil {
		return "", err
	}
	sum := sha256.Sum256([]byte(canonical))
	return base64.RawURLEncoding.EncodeToString(sum[:]), nil
}

func IsEd25519(key JWK) bool {
	return key.Kty == "OKP" && key.Crv == "Ed25519"
}

// PublicKey decodes an Ed25519 public key from its JWK.
func PublicKey(key JWK) (ed25519.PublicKey, error) {
	if !IsEd25519(key) {
		return nil, fmt.Errorf("%w: not an Ed25519 key", ErrKey)
	}
	raw, err := base64.RawURLEncoding.DecodeString(key.X)
	if err != nil {
		return nil, fmt.Errorf("%w: x is not base64url", ErrKey)
	}
	if len(raw) != ed25519.PublicKeySize {
		return nil, fmt.Errorf("%w: x is %d bytes, want %d", ErrKey, len(raw), ed25519.PublicKeySize)
	}
	return ed25519.PublicKey(raw), nil
}

// Validity reports whether a directory key's nbf/exp window contains now, with
// nbf inclusive and exp exclusive.
type Validity string

const (
	ValidityValid       Validity = "valid"
	ValidityNotYetValid Validity = "not-yet-valid"
	ValidityExpired     Validity = "expired"
)

func KeyValidityAt(key JWK, now int64) Validity {
	if key.Nbf != nil && now < *key.Nbf {
		return ValidityNotYetValid
	}
	if key.Exp != nil && now >= *key.Exp {
		return ValidityExpired
	}
	return ValidityValid
}

// PublicOnly reduces any JWK to the members a directory should publish.
//
// A whitelist rather than a blacklist: deleting the private member from an
// exported key leaves key_ops and ext behind, which is a public key advertising
// that it can sign. Enumerating what may be published cannot fail that way.
func PublicOnly(key JWK) JWK {
	return JWK{
		Kty: key.Kty, Crv: key.Crv, X: key.X, Y: key.Y, N: key.N, E: key.E,
		Kid: key.Kid, Alg: key.Alg, Use: key.Use, Nbf: key.Nbf, Exp: key.Exp,
	}
}
